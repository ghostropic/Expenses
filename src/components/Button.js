import React from 'react'
import styled from 'styled-components'

const Button = ({ text, onClick, disabled, className, type }) => (
  // eslint-disable-next-line react/button-has-type
  <button
    type={type}
    disabled={disabled}
    onClick={onClick}
    className={className}
  >
    {text}
  </button>
)

const StyledButton = styled(Button)`
  cursor: pointer;
  height: 48px;
  display: inline-block;
  padding: 0.5rem 0;
  width: 11rem;
  color: ${props => props.color || 'white'};
  border: 2px solid ${props => props.color || 'white'};
  background: ${props => props.background || 'transparent'};

  &.formButton {
    margin: 7px;
  }
  
  &.removeButton {
    background: palevioletred;
  }
 `

export default StyledButton
