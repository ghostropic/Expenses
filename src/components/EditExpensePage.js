import React from 'react'
import styled from 'styled-components'

const EditExpensePage = (props) => {
  console.log(props)
  return (
    <div>
      edit expenses for {props.match.params.id}
    </div>
  )
}

export default EditExpensePage