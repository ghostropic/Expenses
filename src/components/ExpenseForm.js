import React from 'react'
import moment from 'moment'
import {SingleDatePicker} from 'react-dates'

import 'react-dates/initialize'
import 'react-dates/lib/css/_datepicker.css'

export default class ExpenseForm extends React.Component {
  constructor (props) {
    super(props)

    this.state = {
      description: props.expense ? props.expense.description : '',
      amount: props.expense ? (props.expense.amount / 100) : '',
      note: props.expense ? props.expense.note : '',
      createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
      focused: false,
      error: ''
    }
  }

  onDateChange = (createdAt) => {
    if (createdAt) {
      this.setState(() => ({createdAt}))
    }
  }

  onFocusChange = ({focused}) => {
    this.setState(() => ({focused}))
  }

  onDescriptionChange = (e) => {
    const description = e.target.value
    this.setState(() => ({description}))
  }

  onAmountChange = (e) => {
    const amount = e.target.value
    if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
      this.setState(() => ({amount})) 
    }
  }

  onNoteChange = (e) => {
    const note = e.target.value
    this.setState(() => ({note}))
  }

  OnSubmit = (e) => {
    e.preventDefault()
    if (!this.state.description || !this.state.amount) {
      this.setState(() => ({error: 'this is an error message'}))
      console.log('error')
    } else {
      this.setState(() => ({error: ''}))
      this.props.onSubmit({
        description: this.state.description,
        amount: parseFloat(this.state.amount, 10) * 100,
        createdAt: this.state.createdAt.valueOf(),
        note: this.state.note
      })
    }
  }

  render() {
    return (
    <div> 
      <p>Expense Form:</p>
      <div>{this.state.error}</div>
      <form onSubmit={this.OnSubmit}>
        <input 
        value={this.state.description} 
        onChange={this.onDescriptionChange} 
        type="text" 
        placeholder="description" 
        autoFocus
        />
        <input 
        type="text" 
        placeholder="amount"
        value={this.state.amount}
        onChange={this.onAmountChange} 
        /> 
        <textarea 
        placeholder="Add a note to your expense."
        value={this.state.note}
        onChange={this.onNoteChange}
        ></textarea> 
        <SingleDatePicker
          date={this.state.createdAt} // momentPropTypes.momentObj or null
          onDateChange={this.onDateChange} // PropTypes.func.isRequired
          focused={this.state.focused} // PropTypes.bool
          onFocusChange={this.onFocusChange} // PropTypes.func.isRequired
          id="expense_form_calender" // PropTypes.string.isRequired,
          numberOfMonths={1}
          isOutsideRange={() => false}
        />
        <button>Add Expense</button>
      </form>
    </div>
    )
  }
}
