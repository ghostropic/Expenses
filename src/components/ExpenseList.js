import React from 'react'
import { connect } from 'react-redux'

import ExpenseListItem from './ExpenseListItem'
import selectExpenses from '../selectors/expenses'
import selectExpensesTotal from '../selectors/expenses-total'

export const ExpenseList = ({ expenses }) => (
  <div>
    <p>{`Total: ${selectExpensesTotal(expenses)}`}</p>
    <hr />
    {
      expenses.length === 0
        ? (<p>No Expense to list. Try clearing the date range.</p>)
        : (expenses.map(expense => <ExpenseListItem key={expense.id} {...expense} />))
    }
  </div>
)

const MapStateToProps = state => ({
  expenses: selectExpenses(state.expenses, state.filters)
})

export default connect(MapStateToProps)(ExpenseList)
