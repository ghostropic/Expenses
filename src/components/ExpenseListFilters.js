import React from 'react'
import { connect } from 'react-redux'
import { DateRangePicker } from 'react-dates';
import styled from 'styled-components'

import { Input } from '../styles/style'

import {
  setTextFilter,
  sortByDate,
  sortByAmount,
  setStartDate,
  setEndDate
} from '../actions/filters'

class ExpenseListFilters extends React.Component {
  state = {
    calendarFocused: null // string if we are focused on one of the calendars
  }

  onDatesChange = ({ startDate, endDate }) => {
    const { dispatch } = this.props
    dispatch(setStartDate(startDate))
    dispatch(setEndDate(endDate))
  }

  onFocusChange = (calendarFocused) => {
    this.setState(() => ({ calendarFocused }))
  }

  render() {
    const { filters, dispatch } = this.props
    const { calendarFocused } = this.state
    return (
      <FiltersWrapper>
        {/* {console.log('props', this.props)} */}
        <Input
          placeholder=" filter expenses"
          className="normalInput"
          type="text"
          value={filters.text}
          onChange={(e) => {
            dispatch(setTextFilter(e.target.value))
          }}
        />
        <select
          className="normalSelect"
          value={filters.sortBy}
          onChange={(e) => {
            e.target.value === 'date'
              ? dispatch(sortByDate())
              : dispatch(sortByAmount())
          }}
        >
          <option value="amount">Amount</option>
          <option value="date">Date</option>
        </select>
        <DateRangePicker
          startDate={filters.startDate}
          endDate={filters.endDate}
          startDateId="your_unique_start_date_id"
          endDateId="your_unique_end_date_id"
          focusedInput={calendarFocused}
          onFocusChange={this.onFocusChange}
          onDatesChange={this.onDatesChange}
          numberOfMonths={1}
          isOutsideRange={() => false}
          showClearDates
        />
      </FiltersWrapper>
    )
  }
}

const FiltersWrapper = styled.div`
  display: flex;
  align-items: center;
  margin-bottom: 30px;

  select:not([multiple]) {
    -webkit-appearance: none;
    -moz-appearance: none;
    background-position: right 50%;
    background-repeat: no-repeat;
    background-image: url('data:image/svg+xml;utf8,<?xml version="1.0" encoding="utf-8"?><!DOCTYPE svg PUBLIC "-//W3C//DTD SVG 1.1//EN" "http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd"><svg xmlns="http://www.w3.org/2000/svg" width="14" height="12" version="1"><path d="M4 8L0 4h8z"/></svg>');
    padding: .5em;
    padding-right: 1.5em;
  }
  .normalSelect {
    border-radius: 1px;
    border: 0;
  }
  .normalInput, .normalSelect {
    height: 48px;
    border: none;
    margin-right: 7px;
  }
`

const MapStateToProps = state => ({
  filters: state.filters
})

export default connect(MapStateToProps)(ExpenseListFilters)
