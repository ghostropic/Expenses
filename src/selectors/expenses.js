import moment from 'moment'

// getVisibleExpenses - passes expense and filters and determines what should be send along the expense list

export default (expenses, {text, sortBy, startDate, endDate}) => { // two arguments: expenses and destructured filters
  return expenses.filter((expense) => {
    const createdAtMoment = moment(expense.createdAt)
    const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment, 'day') : true
    const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment, 'day') : true
    // const endDateMatch = typeof endDate !== 'number' || expense.createdAt <= endDates
    const textMatch = text === '' || expense.description.toLowerCase().includes(text.toLowerCase())
    return startDateMatch && endDateMatch && textMatch
  }).sort((a,b) => {
    if (sortBy === 'date') {
      return a.createdAt < b.createdAt ? 1 : -1
    } else if (sortBy === 'amount') {
      return a.amount < b.amount ? 1 : -1
    }
  })
  // return [expenses, startDate]
}