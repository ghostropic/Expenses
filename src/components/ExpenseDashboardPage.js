import React from 'react'
import styled from 'styled-components'
import { utils } from '../styles/style'

import ConnectedExpenseList from './ExpenseList';
import ExpenseListFilters from './ExpenseListFilters';

const ExpenseDashboardPage = () => (
  <Dashboard>
    <p>Search expenses.</p>
    <ExpenseListFilters />
    <ConnectedExpenseList />
  </Dashboard>
)

const Dashboard = styled.div`
  padding: ${utils.border};
  max-width: 800px;
`

export default ExpenseDashboardPage
