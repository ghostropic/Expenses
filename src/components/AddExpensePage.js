import React from 'react'
import styled from 'styled-components'
import {connect} from 'react-redux'

import ExpenseForm from './ExpenseForm'
import expenses from '../reducers/expenses'
import { addExpense } from '../actions/expenses'

const AddExpensePage = (props) => (
  <div>
    add expense:
    <ExpenseForm
      onSubmit={(expense) => {
        props.dispatch(addExpense(expense))
        props.history.push('/')
      }}
    />
  </div>
)

export default connect()(AddExpensePage)