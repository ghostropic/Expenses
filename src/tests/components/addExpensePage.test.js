import React from 'react'
import { shallow } from 'enzyme'

import { AddExpensePage } from '../../components/AddExpensePage'
import expenses from '../fixtures/expenses'

let addExpenseFromProps
let history
let wrapper

beforeEach(() => {
  addExpenseFromProps = jest.fn()
  history = { push: jest.fn() }
  wrapper = shallow(<AddExpensePage addExpenseFromProps={addExpenseFromProps} history={history} />)
})

test('should render AddExpensePage correctly', () => {
  expect(wrapper).toMatchSnapshot();
})

test('should handle onSubmit', () => {
  wrapper.find('ExpenseForm').prop('onSubmit')(expenses[1])
  expect(addExpenseFromProps).toHaveBeenLastCalledWith(expenses[1])
  expect(history.push).toHaveBeenCalledWith('/')
})
