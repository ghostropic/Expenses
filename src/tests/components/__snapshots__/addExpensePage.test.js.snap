// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render AddExpensePage correctly 1`] = `
<AddExpensePage__AddExpensePageWrapper>
  <p>
    Add an expense.
  </p>
  <ExpenseForm
    onSubmit={[Function]}
  />
</AddExpensePage__AddExpensePageWrapper>
`;
